qutip/solver/ode/explicit_rk.pyx

```diff
@@ -0,0 +1,485 @@
+                                              self._y_temp, <Data> self.k[i])
+
+        # Compute the state
+        self._y_front = copy_to(self._y_prev, self._y_front)
+        self._y_front = self._accumulate(self._y_front, self.b, dt, self.rk_step)
+        self._t_front = self._t_prev + dt
+
+        if type(self._y_front) is CSR:
+            # issparse() test would be better.
+            tidyup_csr(self._y_front, self.atol/self._y_front.shape[0], True)
+
+        return self._error(self._y_front, dt)
+
+    cdef double _error(self, Data y_new, double dt):
+        """ Compute the normalized error. (error/tol) """
+        if not self.adaptative_step:
+            return 0.
+        imul_data(self._y_temp, 0.)
+        self._y_temp = self._accumulate(self._y_temp, self.e, dt, self.rk_step)
+        self._norm_front = frobenius_data(y_new)
+        return frobenius_data(self._y_temp) / (self.atol +
+                max(self._norm_prev, self._norm_front) * self.rtol)
+
+    cdef void _prep_dense_out(self):
+        """
+        Compute derivative for the interpolation step.
+        """
+        cdef double dt = self._dt_int
+
+        for i in range(self.rk_step, self.rk_extra_step):
+            imul_data(<Data> self.k[i], 0.)
+            self._y_temp = copy_to(self._y_prev, self._y_temp)
+            self._y_temp = self._accumulate(self._y_temp, self.a[i,:], dt, i)
+            self.k[i] = self.qevo.matmul_data(self._t_prev + self.c[i]*dt,
+                                              self._y_temp, <Data> self.k[i])
+
+    cdef Data _interpolate_step(self, double t, Data out):
+        """
+        Compute the state at any points between _t_prev and _t_front
+        with an error of ~dt**denseout_order.
+        """
+        cdef:
+            int i, j
+            double t0 = self._t_prev
+            double dt = self._dt_int
+            double tau = (t - t0) / dt
+        for i in range(self.rk_extra_step):
+            self.b_factor[i] = 0.
+            for j in range(self.denseout_order-1, -1, -1):
+                self.b_factor[i] += self.bi[i,j]
+                self.b_factor[i] *= tau
+
+        out = copy_to(self._y_prev, out)
+        out = self._accumulate(out, self.b_factor, dt, self.rk_extra_step)
+        return out
+
+    cdef inline Data _accumulate(self, Data target, double[:] factors,
+                         double dt, int size):
+        cdef int i
+        for i in range(size):
+            target = iadd_data(target, <Data> self.k[i], dt * factors[i])
+        return target
+
+    @cython.cdivision(True)
+    cdef double _get_timestep(self, double t):
+        """ Get the dt for the step. """
+        cdef double dt_needed = t - self._t_front
+        if not self.adaptative_step:
+            return dt_needed
+        if self.interpolate:
+            return self._dt_safe
+        if dt_needed <= self._dt_safe:
+            return dt_needed
+        return dt_needed / (int(dt_needed / self._dt_safe) + 1)
```

Comment:
I don't understand why we don't just return `self._dt_safe` at this point? This formula is almost the same as `self._dt_safe` but I'm guessing it is doing something I am not seeing?